src/Toml/Parser/Value.hs

```diff
@@ -170,10 +171,23 @@ doubleP :: Parser Double
 doubleP = lexeme (signed sc (num <|> inf <|> nan)) <?> "double"
   where
     num, inf, nan :: Parser Double
-    num = float
+    num = floatP
     inf = 1 / 0 <$ string "inf"
     nan = 0 / 0 <$ string "nan"
 
+floatP :: Parser Double
+floatP = check =<< readMaybe . concat <$> sequence [ digits, expo <|> dot ]
+  where
+    check = maybe (fail "Not a float") return
+
+    digits, dot, expo :: Parser String
+    digits = concat <$> sepBy1 (some digitChar) (char '_')
+    dot = concat <$> sequence [pure <$> char '.', digits, option "" expo]
+    expo = concat <$> sequence
```

Comment:
I feel like you can implement helper function for `concat <$> sequence`. And then maybe we can notice that such function already exist.